go/vt/vtgate/tabletgateway.go

```diff
@@ -231,10 +258,22 @@ func (gw *TabletGateway) withRetry(ctx context.Context, target *querypb.Target,
 				defer retryDone()
 				bufferedOnce = true
 			}
+
+			if bufferErr != nil {
+				err = vterrors.Wrapf(bufferErr,
+					"failed to automatically buffer and retry failed request during failover. original err (type=%T): %v",
+					err, err)
+				break
+			}
 		}
 
 		tablets := gw.hc.GetHealthyTabletStats(target)
 		if len(tablets) == 0 {
+			// if we have a keyspace event watcher, check if the reason why our primary is not available is that it's currently being resharded
+			if target.TabletType == topodatapb.TabletType_PRIMARY && gw.kev != nil && gw.kev.TargetIsBeingResharded(target.Keyspace, target.Shard) {
+				err = vterrors.Errorf(vtrpcpb.Code_CLUSTER_EVENT, "current keyspace is being resharded")
+				continue
+			}
```

Comment:
This is the special handling I was talking about.

Reply:
Question to make sure I'm following correctly this change: Now, in addition to previous existent logic, it will use these events to retry. 

One follow up question for these retries: other errors in this block, were not that time sensitive. For instance, if a connection to a tablet failed, in the next iteration of the block another tablet would be used for the retry. In this case, there is some amount of time that needs to elapsed before new primaries will become available.  Would it make sense to start thinking of exponential backoff between retries?

Reply:
There is no need for exponential backoff here. In this specific case, we're setting an explicit buffering error, so that the next iteration of the loop will block directly on the buffering code and will not resume until the primary becomes available. This is not a busy loop, so we don't need to backoff. :ok_hand:

Reply:
Nice, this makes sense.

Reply:
Am I right that the "buffer" request here is not really put a "request" into a queue, we are queueing the entry and the actaul request goroutine just gets block until the shard split/failover finished?